plugins/relengapi/relengapi.go

```diff
@@ -0,0 +1,158 @@
+package relengapi
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httputil"
+	"time"
+
+	"github.com/pkg/errors"
+	schematypes "github.com/taskcluster/go-schematypes"
+	"github.com/taskcluster/taskcluster-worker/engines"
+	"github.com/taskcluster/taskcluster-worker/plugins"
+	"github.com/taskcluster/taskcluster-worker/runtime"
+)
+
+const tmpTokenLifetime time.Duration = 10 * time.Minute
+const tmpTokenSkew time.Duration = 10 * time.Second
+
+type provider struct {
+	plugins.PluginProviderBase
+}
+
+type plugin struct {
+	plugins.PluginBase
+	config config
+}
+
+type taskPlugin struct {
+	plugins.TaskPluginBase
+	monitor           runtime.Monitor
+	proxy             httputil.ReverseProxy
+	context           *runtime.TaskContext
+	tmpToken          string
+	tmpTokenGoodUntil time.Time
+	permissions       []string
+	relengapiHost     string
+}
+
+func init() {
+	plugins.Register("relengapi", provider{})
+}
+
+func (provider) ConfigSchema() schematypes.Schema {
+	return configSchema
+}
+
+func (p *plugin) PayloadSchema() schematypes.Object {
+	return payloadSchema
+}
+
+func (provider) NewPlugin(options plugins.PluginOptions) (plugins.Plugin, error) {
+	var c config
+	schematypes.MustValidateAndMap(configSchema, options.Config, &c)
+	if c.Host == "" {
+		c.Host = "api.pub.build.mozilla.org"
+	}
+
+	if c.Token == "" {
+		err := errors.New("Token value not configured for relengapi plugin")
+		debug(fmt.Sprintln(err))
+		options.Monitor.Error(err)
+		return nil, err
+	}
+
+	return &plugin{
+		PluginBase: plugins.PluginBase{},
+		config:     c,
+	}, nil
+}
+
+func (p *plugin) NewTaskPlugin(options plugins.TaskPluginOptions) (plugins.TaskPlugin, error) {
+	var P payload
+	schematypes.MustValidateAndMap(payloadSchema, options.Payload, &P)
+
+	// If disabled, we return nothing
+	if !P.EnableRelengAPIProxy {
+		return plugins.TaskPluginBase{}, nil
+	}
+
+	debug(fmt.Sprintf("Starting relengapi plugin for task %s", options.TaskContext.TaskID))
+	q := options.TaskContext.Queue()
+	task, err := q.Task(options.TaskContext.TaskID)
```

Comment:
Instead of doing this we could expose `Scopes` from `TaskInfo` instead. But there is a very good reason we don't do this. Scopes should never be read as a list.



There is only one question you can ask: Does `task.scopes` satisfy "X"?
https://github.com/taskcluster/taskcluster-worker/blob/master/runtime/taskcontext.go#L306

A better option is something like:
```go
allPerms := []string{
  "base.tokens.tmp.issue",
  "base.tokens.usr.issue",
  "base.tokens.usr.revoke.my",
  "base.tokens.usr.view.my",
  "clobberer.post.clobber",
  "tooltool.download.internal",
  "tooltool.download.public",
  "tooltool.upload.internal",
  "tooltool.upload.public",
}
perms := []string{}
for _, p := range allPerms {
  if options.TaskContext.HasScopes("worker:relengapi-proxy:" + p) {
    perms = append(perms, p)
  }
}
```

I'm not sure there is an exhaustive list of all permissions. But just adding the ones we think we need is a good start.

Reply:
Note: extracting information from scopes is always wrong. In this case we ensure that kleene works as expected, because `task.scopes = ['worker:relengapi-proxy:*']` will have all permissions, similarly `task.scopes = ['w*']` will also have all permissions.